### DISCUSSION TITLE
Mention py_modules in "Package Discovery and Namespace Package"?

### DISCUSSION THREAD

**Discussion by kinow** [pypa/setuptools#3288] [Q&A]
Hi,

I started using `pyproject.toml` and `setup.cfg` today after a colleague [told me](https://github.com/common-workflow-lab/wdl-cwl-translator/pull/196#discussion_r858877748) about it.

I had an AWS Lambda script that I also run locally, in a repository with `requirements.txt` files and the `app.py` single script. Instead of migrating to `setup.py`, I went straight to `pyproject.toml` and tried to use the [Package Discovery and Namespace Package](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#single-module-distribution) docs, specifically the section “single-module distribution”.

Its description matches, I think, exactly my case,

>A standalone module is placed directly under the project root, instead of inside a package folder:

I tried first without anything, letting the build tools to try to find my `app.py` (right next to `pyproject.toml` and `setup.cfg), but that didn't work (it failed to include the `app.py` in the produced sdist file).

After some googling, I found that this fixed the issue for me:

```py
[options]
py_modules =
    app
```

Is it something that would be worth an example for that in the “Package Discovery and Namespace Package” page? I found examples for src-layout, for the flat-layout, but the variation of a single module project appears to need `py_modules`, which may not be intuitive I guess.

Thanks!
Bruno

**Comment by abravalheri** [comment]
Hi @kinow, thanks for reaching out.

Would you like to propose a change in the docs to target this concern? We are always welcoming PRs :)

> I tried first without anything, letting the build tools to try to find my `app.py`

This seems to be a bit off... The `app.py` module should be picked up, unless you are already setting `packages`...

For example:

```bash
cd /tmp
rm -rf /tmp/myproj
mkdir /tmp/myproj
touch /tmp/myproj/app.py
cat <<EOS > /tmp/myproj/pyproject.toml
[build-system]
requires = ['setuptools>=62.1.0']
build-backend = 'setuptools.build_meta'

[project]
name = "myproj"
version = "42"
EOS
pip-run build -- -m build /tmp/myproj
unzip -l /tmp/myproj/dist/*.whl
```

Should give us:

```console
Successfully built myproj-42.tar.gz and myproj-42-py3-none-any.whl
Archive:  /tmp/myproj/dist/myproj-42-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2022-04-27 10:42   app.py
      109  2022-04-27 10:43   myproj-42.dist-info/METADATA
       92  2022-04-27 10:43   myproj-42.dist-info/WHEEL
        4  2022-04-27 10:43   myproj-42.dist-info/top_level.txt
      340  2022-04-27 10:43   myproj-42.dist-info/RECORD
---------                     -------
      545                     5 files
```

**Reply by kinow** [reply]
That worked! That's funny, I thought I had tested without the `packages` but I left this part:

```
package_dir =
    =.*
```

Removing that, and also the py_modules` fixed it.

But there's something else (sorry if I'm digressing). I created a tiny project to test it, using my other project and your example as reference, to compare the settings. I've put the contents in this gist: https://gist.github.com/kinow/037ea9c5175d715942532b2a801da1c5

I was getting a `ModuleNotFoundError: app` every time I tried `python -m build`. That `package_dir` is invalid, and supposed to raise an error, which it does, but only when I don't reference `app.py` in the `[metadata]version`. See comment in the gist.

I think if it raised the same error in both cases, that would make more sense. I also had a `[options.entry_points]`, which I thought was causing the `ModuleNotFoundError`, but only found it was the `version` after slowly adding items using your example. Not sure if it's a bug, or expected behavior.

>Would you like to propose a change in the docs to target this concern? We are always welcoming PRs :)

That's what I thought would be nice to have, but now since the automatic configuration works, as long as you don't have `packages` nor `package_dir`, I'm not sure if that's necessary.

Thanks a lot @abravalheri !
Bruno
